src/IO/StackDataReader.cs

```diff
@@ -34,77 +47,247 @@ public StackDataReader(byte[] data, long len)
+
+            Skip(8);
+
+            return v;
+        }
+
+
+
+
+
+        [MethodImpl(IMPL_OPTION)]
+        public ushort ReadUInt16BE()
+        {
+            ushort v = (ushort)((Data[Position] << 8) |
+                                 (Data[Position + 1]));
+
+            Skip(2);
 
-                if (c != 0)
+            return v;
+        }
+
+        [MethodImpl(IMPL_OPTION)]
+        public short ReadInt16BE()
+        {
+            short v = (short)((Data[Position] << 8) |
+                              (Data[Position + 1]));
+
+            Skip(2);
+
+            return v;
+        }
+
+        [MethodImpl(IMPL_OPTION)]
+        public uint ReadUInt32BE()
+        {
+            uint v = (uint)((Data[Position] << 24) |
+                            (Data[Position + 1] << 16) |
+                            (Data[Position + 2] << 8) |
+                            (Data[Position + 3]));
+
+            Skip(4);
+
+            return v;
+        }
+
+        [MethodImpl(IMPL_OPTION)]
+        public int ReadInt32BE()
+        {
+            int v = (int)((Data[Position] << 24) |
+                          (Data[Position + 1] << 16) |
+                          (Data[Position + 2] << 8) |
+                          (Data[Position + 3]));
+
+            Skip(4);
+
+            return v;
+        }
+
+        [MethodImpl(IMPL_OPTION)]
+        public ulong ReadUInt64BE()
+        {
+            ulong v = (ulong)(((ulong)Data[Position] << 56) |
+                              ((ulong)Data[Position + 1] << 48) |
+                              ((ulong)Data[Position + 2] << 40) |
+                              ((ulong)Data[Position + 3] << 32) |
+                              ((ulong)Data[Position + 4] << 24) |
+                              ((ulong)Data[Position + 5] << 16) |
+                              ((ulong)Data[Position + 6] << 8) |
+                              ((ulong)Data[Position + 7]));
+
+            Skip(8);
+
+            return v;
+        }
+
+        [MethodImpl(IMPL_OPTION)]
+        public long ReadInt64BE()
+        {
+            long v = (long)(((ulong)Data[Position] << 56) |
+                              ((ulong)Data[Position + 1] << 48) |
+                              ((ulong)Data[Position + 2] << 40) |
+                              ((ulong)Data[Position + 3] << 32) |
+                              ((ulong)Data[Position + 4] << 24) |
+                              ((ulong)Data[Position + 5] << 16) |
+                              ((ulong)Data[Position + 6] << 8) |
+                              ((ulong)Data[Position + 7]));
+
+            Skip(8);
+
+            return v;
+        }
+
+
+
+        [MethodImpl(IMPL_OPTION)]
+        public T* CastTo<T>() where T : unmanaged
+        {
+            return (T*)PositionAddress;
+        }
+
+        public string ReadASCII(int size)
+        {
+            ValueStringBuilder sb = new ValueStringBuilder(size);
```

Comment:
I think it is faster to use `Encoding.ASCII.GetString()` because the conversion is vectorized. We should benchmark to see.
But that might be a .NET Core/5 optimization, not .NET Framework. 🤔

Reply:
From my experience i met some server that sends garbage inside the string like:
` 'C', 'L', 'A', 'S', 'S', 'I', 'C', '\0', 'U', 'O' ` but the original client understand it as `CLASSICUO`.
Using `Encoding.ASCII.GetString()` will cut the string at '\0'

Reply:
This is true, and that is due to fixed size vs non-fixed size IIRC. If that is the case, you would need to remove all invalid chars, otherwise you can use ASCII get string.